tests/events_test.go

```diff
@@ -0,0 +1,206 @@
+				AccessModes:      []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+				PersistentVolumeSource: corev1.PersistentVolumeSource{
+					Local: &corev1.LocalVolumeSource{
+						Path: "/dev/mapper/" + deviceName,
+					},
+				},
+			},
+		}
+		pvc := &corev1.PersistentVolumeClaim{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: ioerrorPVC,
+			},
+			Spec: corev1.PersistentVolumeClaimSpec{
+				VolumeMode:       &vMode,
+				StorageClassName: &sc,
+				AccessModes:      []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+				Resources: corev1.ResourceRequirements{
+					Requests: map[corev1.ResourceName]resource.Quantity{corev1.ResourceStorage: size},
+				},
+			},
+		}
+		virtCli, err := kubecli.GetKubevirtClient()
+		Expect(err).ToNot(HaveOccurred())
+
+		_, err = virtCli.CoreV1().PersistentVolumes().Create(context.Background(), pv, metav1.CreateOptions{})
+		if !errors.IsAlreadyExists(err) {
+			Expect(err).ToNot(HaveOccurred())
+		}
+
+		_, err = virtCli.CoreV1().PersistentVolumeClaims(ns).Create(context.Background(), pvc, metav1.CreateOptions{})
+		if !errors.IsAlreadyExists(err) {
+			Expect(err).ToNot(HaveOccurred())
+		}
+	}
+
+	createVMIwithFaultyPVC := func(ns string) {
+		vmi = tests.NewRandomVMIWithNS(ns)
+		bus := "virtio"
+		vmi.Spec.Domain.Devices.Disks = append(vmi.Spec.Domain.Devices.Disks, v1.Disk{
+			Name: diskName,
+			DiskDevice: v1.DiskDevice{
+				Disk: &v1.DiskTarget{
+					Bus: bus,
+				},
+			},
+		})
+		vmi.Spec.Volumes = append(vmi.Spec.Volumes, v1.Volume{
+			Name: diskName,
+			VolumeSource: v1.VolumeSource{
+				PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+					ClaimName: ioerrorPVC,
+				},
+			},
+		})
+
+		vmi.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory] = resource.MustParse("512M")
+		_, err := virtClient.VirtualMachineInstance(ns).Create(vmi)
+		Expect(err).To(BeNil(), "Create VMI successfully")
+	}
+
+	removeVMI := func(ns string) {
+		err := virtClient.VirtualMachineInstance(ns).Delete(vmi.ObjectMeta.Name, &metav1.DeleteOptions{})
+		Expect(err).To(BeNil(), "Delete VMI successfully")
+	}
+
+	removePVwithFaultyDisk := func() {
+		err := virtClient.CoreV1().PersistentVolumes().Delete(context.Background(), ioerrorPV, metav1.DeleteOptions{})
+		Expect(err).ToNot(HaveOccurred())
+	}
+
+	isExpectedIOEvent := func(e corev1.Event) bool {
+		if e.Type == "Warning" &&
+			e.Reason == "IOerror" &&
+			e.Message == "VM Paused due to IO error at the volume: "+diskName &&
+			e.InvolvedObject.Kind == "VirtualMachineInstance" &&
+			e.InvolvedObject.Name == vmi.ObjectMeta.Name {
+			return true
+		}
+		return false
+	}
+
+	BeforeEach(func() {
+		ns = tests.NamespaceTestDefault
+		createFaultyDisk()
+		createPVCwithFaultyDisk(ns)
+	})
+	AfterEach(func() {
+		removeVMI(ns)
+		removePVwithFaultyDisk()
+		// Before being able to remove the disk, the vmi has to be terminated
+		tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)
+		removeFaultyDisk()
```

Comment:
Since this `removeFaultyDisk` is mutating a host node, we really want this to happen even if the VMI and PVCs aren't removed properly. 

Can we still cleanup properly with this `removeFaultyDisk` function being the first thing that's executed in AfterEach?

Reply:
@davidvossel yes, I understand your concerns. Unfortunately, if the disk is still in use by the VM, the dmsetup returns `resource busy` and fails. I'll find a way to trigger the cleanup

Reply:
I moved the removal of the vmi in the test and added an eventually to loop a couple of time to remove the disk. Now, the `removeFaultyDisk` is the first action that is performed in the AfterEach